### PR TITLE
First take on Graph Generators

### PR DESCRIPTION
**Pull request by miratepuffin**

### CODE REVIEW

File: docbrown/db/src/graphgen/preferential_attachment.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        graph.add_vertex(latest_time,0,&vec![]);
+        degrees.push(0);
+        ids.push(0);
+    }
+    while ids.len() < edges_per_step {
+        max_id+=1;
+        latest_time+=1;
+        let prev_node = ids[ids.len()-1];
+        graph.add_vertex(latest_time,max_id,&vec![]);
+        graph.add_edge(latest_time,max_id,prev_node,&vec![]);
```
Comment (ljeub-pometry) [inline docbrown/db/src/graphgen/preferential_attachment.rs:33]: Shouldn't we add the edge in both directions?

File: docbrown/db/src/graphgen/preferential_attachment.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            let mut sum = 0;
+            let rand_num = rng.gen_range(1..=normalisation);
+            for(pos,id) in ids.iter().enumerate() {
+                if ! positions_to_skip.contains(&pos){
+                    sum += degrees[pos];
+                    if sum>= rand_num {
+                        positions_to_skip.insert(pos);
+                        normalisation-degrees[pos];
+                        break;
+                    }
```
Comment (ljeub-pometry) [inline docbrown/db/src/graphgen/preferential_attachment.rs:58]: Not sure this is the most efficient sampler though it does look ok. I suppose the structure of the preferential attachment helps a lot here to keep this somewhat efficient as nodes with large degree should be near the beginning of the list...

File: docbrown/db/src/graphgen/preferential_attachment.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                }
+
+            }
+        }
+        for pos in positions_to_skip {
+            let dst = ids[pos];
+            degrees[pos]+=1;
+            graph.add_edge(latest_time,max_id,dst,&vec![]);
```
Comment (ljeub-pometry) [inline docbrown/db/src/graphgen/preferential_attachment.rs:66]: Probably should add both edge directions

File: docbrown/db/src/graphgen/random_attachment.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+use crate::graph::Graph;
+use rand::seq::SliceRandom;
+
+pub fn random_attachment(graph:&Graph, vertices_to_add:usize,edges_per_step:usize) {
```
Comment (ljeub-pometry) [inline docbrown/db/src/graphgen/random_attachment.rs:4]: Add some doc comments

File: docbrown/db/src/graphgen/random_attachment.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        graph.add_vertex(latest_time,max_id,&vec![]);
+        ids.push(max_id);
+    }
+
+    for i in 0..vertices_to_add {
+        let edges = ids.choose_multiple(rng,edges_per_step);
+        max_id+=1;
+        latest_time+=1;
+        edges.for_each(|neighbour| {
+            graph.add_edge(latest_time, max_id, *neighbour, &vec![]);
```
Comment (ljeub-pometry) [inline docbrown/db/src/graphgen/random_attachment.rs:26]: add both directions?

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
- earliest/latest time should return options
- probably should be making undirected networks

File: docbrown/db/src/graphgen/preferential_attachment.rs (hunk 1 above)
Comment (miratepuffin) [reply docbrown/db/src/graphgen/preferential_attachment.rs:33]: May thinking with this is that with the undirected views etc a user wanting the edge in both ways could just utilise that.

File: docbrown/db/src/graphgen/preferential_attachment.rs (hunk 2 above)
Comment (miratepuffin) [reply docbrown/db/src/graphgen/preferential_attachment.rs:58]: I am sure there is a better way of doing this out there, however, there doesn't seem to be any decent data structures for maintaining a weighted index for sampling. For those that do exist, you can only update the weights not add new elements to the structure without rebuilding. These often don't allow for sampling without replacement as well which is key here.

**Pull request merged**